dietpi/dietpi-software

```diff
@@ -9018,7 +9018,7 @@ _EOF_
 			
 			if [[ ! -f '/opt/AdGuardHome/AdGuardHome.yaml' ]]; then
 				dps_index=$software_id Download_Install 'AdGuardHome.yaml' /opt/AdGuardHome/AdGuardHome.yaml
-				G_CONFIG_INJECT 'password:[[:blank:]]' "  password: $(htpasswd -bnBC 10 "" $GLOBAL_PW | tr -d ':\n' | sed 's/$2y/$2a/')" /opt/AdGuardHome/AdGuardHome.yaml
+				G_CONFIG_INJECT 'password:[[:blank:]]' "  password: $(htpasswd -bnBC 10 "" $GLOBAL_PW | tr -d ":\n" | sed "s/$2y/$2a/")" /opt/AdGuardHome/AdGuardHome.yaml
```

Comment:
Hehe that fix was wrong. Now the shell tries to expand `$2`, which was never set, hence two empty strings are passed there. The single quotes were perfectly fine, which tell the shell to interpret all contained characters literally. To address the shellcheck warning and also avoid that `sed` interprets `$` as "end of string" (special regular expressions meaning, although in this case `sed` is smart enough), you can escape it with a backslash to tell everyone that it is meant literal:
```
G_CONFIG_INJECT 'password:[[:blank:]]' "  password: $(htpasswd -bnBC 10 '' '$GLOBAL_PW' | tr -d ':\n' | sed 's/\$2y/\$2a/')" /opt/AdGuardHome/AdGuardHome.yaml
```
The single quotes around `$GLOBAL_PW` are important to avoid issues with special characters and spaces in the password. The variable is still expanded as the whole thing is double quoted. So the shell first expands `$GLOBAL_PW` and then passes it together with the single quotes to the `$(...)` command substitution which hence has a single-quoted literal password string. The more I write to more I see that it's quite complicated with nested commands, command substitutions and commands which all have their syntax and special character interpretation rules 😄.

Reply:
well this is not working as well 😉 
single quotes around $`GLOBAL_PW` will set `$GLOBAL_PW` as password 🤣

Reply:
An lol, true, as it's inside of the command substitution. Okay then use double quotes, around it.

Reply:
😟 still complaining

Reply:
I disabled the check for this line now. I though I solved it in other cases inside `sed` with a backslash already 🤔: https://github.com/koalaman/shellcheck/wiki/SC2016
> In the particular case of `sed`, ShellCheck uses additional heuristics to try to separate cases like `'s/$foo/bar/'` (failing to replace the variable `$foo`) with from the false positives like `'$d'` (delete last line).

Not so smart, as a backslash is THE sign for "take it literal" 😉. I'll open an issue there, when I find time to search through all the existing SC2016 false alarm reports that exist already 😄.